opentracing-api/src/main/java/io/opentracing/io/opentracing/mocktracer/MockTracer.java

```diff
@@ -0,0 +1,160 @@
+package io.opentracing.io.opentracing.mocktracer;
+
+import io.opentracing.References;
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.propagation.Format;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * MockTracer makes it easy to test the semantics of OpenTracing instrumentation.
+ *
+ * By using a MockTracer as an io.opentracing.Tracer implementation for unittests, a developer can assert that Span
+ * properties and relationships with other Spans are defined as expected by instrumentation code.
+ *
+ * The MockTracerTest has simple usage examples.
+ */
+public class MockTracer implements Tracer {
+    private List<MockSpan> finishedSpans = new ArrayList<>();
+    private final Propagator propagator;
+
+    public MockTracer() {
+        this(Propagator.PRINTER);
+    }
+
+    /**
+     * Create a new MockTracer that passes through any calls to inject() and/or extract().
+     *
+     * @param propagator
+     */
+    public MockTracer(Propagator propagator) {
+        this.propagator = propagator;
+    }
+
+    /**
+     * Clear the finishedSpans() queue.
+     *
+     * Note that this does *not* have any effect on Spans created by MockTracer that have not finish()ed yet; those
+     * will still be enqueued in finishedSpans() when they finish().
+     */
+    public synchronized void reset() {
+        this.finishedSpans.clear();
+    }
+
+    /**
+     * @return a copy of all finish()ed MockSpans started by this MockTracer (since construction or the last call to
+     * MockTracer.reset()).
+     *
+     * @see MockTracer#reset()
+     */
+    public synchronized List<MockSpan> finishedSpans() {
+        return new ArrayList<>(this.finishedSpans);
+    }
+
+    /**
+     * Propagator allows the developer to intercept and verify any calls to inject() and/or extract().
+     *
+     * By default, MockTracer uses Propagator.PRINTER which simply logs such calls to System.out.
+     *
+     * @see MockTracer#MockTracer(Propagator)
+     */
+    public interface Propagator {
+        <C> void inject(MockSpan.MockContext ctx, Format<C> format, C carrier);
+        <C> MockSpan.MockContext extract(Format<C> format, C carrier);
+
+        static Propagator PRINTER = new Propagator() {
```

Comment:
having done quite a lot of instrumentation of RPC frameworks, I found the most reliable test to be a client-server pair that actually exchange network messages, and validate that tracing is correctly propagated. That kind of test requires a real propagator, similar to MockTracer in Go.